### PR TITLE
Add support for GOAWAY frame

### PR DESCRIPTION
**Pull request by duke8253**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

Add support for GOAWAY frame, also adjusted the RST_STREAM frame autests.

### CODE REVIEW

File: Sconstruct [hunk 1]
```diff
@@ -113,19 +113,9 @@ else:
...
-                        # author rebases origin/quic.  For reference, this
-                        # commit is currently described as:
-
-                        # commit f277dafe1ef946dac8d7bd57ac4d30e722b654d9 (HEAD -> master, origin/master, origin/HEAD)
-                        # Author: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
-                        # Date:   Sun Apr 3 21:56:47 2022 +0900
-                        #
-                        #     Bump macos
-                        revision="f277dafe1ef946dac8d7bd57ac4d30e722b654d9"
+                        tag="v1.51.0"
```
Comment (bneradt) [inline Sconstruct:118]: Thanks for helping keep these updated.

File: local/src/core/http2.cc [hunk 2]
```diff
@@ -731,35 +736,61 @@ on_frame_send_cb(nghttp2_session *session, nghttp2_frame const *frame, void *use
...
+    } else if (frame->hd.type == goaway_after) {
+      errata.note(
+          S_DIAG,
+          "Submitting GOAWAY frame for key {} after {} frame with error code {}.",
+          stream_state._key,
+          H2FrameNames[static_cast<H2Frame>(goaway_after)],
+          H2ErrorCodeNames[static_cast<H2ErrorCode>(goaway_error)]);
+      errata.sink();
+      int const submit_status = nghttp2_session_terminate_session2(session, 0, goaway_error);
+      if (submit_status) {
```
Comment (bneradt) [inline local/src/core/http2.cc:784]: Nit: can you make this `submit_status != 0`. Only with boolean vars do we not compare to some value.

File: local/src/core/http2.cc [hunk 3]
```diff
@@ -780,8 +811,10 @@ on_frame_recv_cb(nghttp2_session * /* session */, nghttp2_frame const *frame, vo
     break;
   case NGHTTP2_PRIORITY:
     break;
-  case NGHTTP2_RST_STREAM:
-    break;
+  case NGHTTP2_RST_STREAM: {
+    // nghttp2_rst_stream const &rst_stream_frame{frame->rst_stream};
+    errata.note(S_DIAG, "Received RST_STREAM frame with stream id {}", frame->hd.stream_id);
```
Comment (bneradt) [inline local/src/core/http2.cc:816]: While you're at it, can you also add the error code to this? That could be helpful under some circumstances.

File: local/src/core/http2.cc [hunk 4]
```diff
@@ -800,8 +833,13 @@ on_frame_recv_cb(nghttp2_session * /* session */, nghttp2_frame const *frame, vo
...
   case NGHTTP2_PUSH_PROMISE:
     break;
-  case NGHTTP2_GOAWAY:
-    break;
+  case NGHTTP2_GOAWAY: {
+    // nghttp2_goaway const &goaway_frame{frame->goaway};
+    errata.note(
+        S_DIAG,
+        "Received GOAWAY frame with last stream id {}",
+        frame->goaway.last_stream_id);
```
Comment (bneradt) [inline local/src/core/http2.cc:841]: Error code here too please.

File: test/autests/gold_tests/http2_abort/replay_files/client_goaway_after_headers.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
+meta:
```
Comment (bneradt) [inline test/autests/gold_tests/http2_abort/replay_files/client_goaway_after_headers.yaml:1]: We should add the Apache license for replay files. I think I've missed that in other replay files... we should add that with a separate commit. Don't worry for this PR. But for this new file, let's add the license.

**Review by bneradt — CHANGES_REQUESTED** [review]
Looks great to me. Just some minor observations. Thanks for working on this.

File: README.md [hunk 6]
```diff
@@ -484,6 +486,12 @@ The available options for `error-code` are:
 * INADEQUATE_SECURITY
 * HTTP_1_1_REQUIRED
 
+#### GOAWAY frame
+
+The `GOAWAY` frame acts similar to the `RST_STREAM` frame shown above. But rather than terminating
+a stream, `GOAWAY` frame terminates the connection. It also uses the same set of `error-code` listed
```
[the remark this answers is not in the document]
Comment (duke8253) [reply README.md:492]: But the RFC doc says this 

> The GOAWAY frame (type=0x7) is used to initiate shutdown of a connection or to signal serious error conditions.

[the remark this answers is not in the document]
Comment (bneradt) [reply README.md:492]: Good point. Let's stick with the paragraph as you wrote it. Thanks for pointing this out.

**Review by bneradt — APPROVED** [review]
Looks good. Thanks for the quick turn around on reviews.

**Pull request merged**